lib/icinga/apiactions.cpp

```diff
@@ -555,5 +556,164 @@ Dictionary::Ptr ApiActions::GenerateTicket(const ConfigObject::Ptr&,
+
+	String endpoint = "$command_endpoint$";
+	if (params->Contains("endpoint"))
+		endpoint = params->Get("endpoint");
+
+	MacroProcessor::ResolverList resolvers;
+	if (params->Contains("macros")) {
+		Value macros = HttpUtility::GetLastParameter(params, "macros");
+		if (macros.IsObjectType<Dictionary>())
+			resolvers.emplace_back("override", macros);
+		else
+			return ApiActions::CreateResult(400, "Parameter macros must be a dictionary.");
+	}
+
+	if (service)
+		resolvers.emplace_back("service", service);
+	resolvers.emplace_back("host", host);
+	resolvers.emplace_back("icinga", IcingaApplication::GetInstance());
+
+	String resolved_endpoint = MacroProcessor::ResolveMacros(
+		endpoint, resolvers, checkable->GetLastCheckResult(),
+		nullptr, MacroProcessor::EscapeCallback(), nullptr, false
+	);
+
+	/* Check if endpoint exists */
+	Endpoint::Ptr endpointPtr = Endpoint::GetByName(resolved_endpoint);
+	if (!endpointPtr)
+		return ApiActions::CreateResult(404, "Can't find a valid endpoint for '" + resolved_endpoint + "'.");
+
+	/* Get command_type */
+	String command_type = "EventCommand";
+	if (params->Contains("command_type"))
+		command_type = HttpUtility::GetLastParameter(params, "command_type");
+
+	/* Validate command_type */
+	if (command_type != "EventCommand" && command_type != "CheckCommand" && command_type != "NotificationCommand")
+		return ApiActions::CreateResult(400, "Invalid command_type '" + command_type + "'.");
+
+	/* Get command */
+	String command;
+	if (!params->Contains("command")) {
+		if (command_type == "CheckCommand" ) {
+			command = "$check_command$";
+		} else if (command_type == "EventCommand") {
+			command = "$event_command$";
+		} else if (command_type == "NotificationCommand") {
+			command = "$notification_command$";
+		}
+	} else {
+		command = HttpUtility::GetLastParameter(params, "command");
+	}
+
+	/* Resolve command macro */
+	String resolved_command = MacroProcessor::ResolveMacros(
+		command, resolvers, checkable->GetLastCheckResult(), nullptr,
+		MacroProcessor::EscapeCallback(), nullptr, false
+	);
+
+	/* Check if resolved_command is not empty */
+	if (resolved_command.IsEmpty())
+		return ApiActions::CreateResult(400, "Command must not be empty.");
+
+	/* Check if resolved_command exists and it is of type command_type */
+	if (command_type == "CheckCommand") {
+		if (!CheckCommand::GetByName(resolved_command))
+			return ApiActions::CreateResult(400, "Command '" + resolved_command + "' is not of type '" + command_type + "'.");
+	} else if (command_type == "EventCommand") {
+		if (!EventCommand::GetByName(resolved_command))
+			return ApiActions::CreateResult(400, "Command '" + resolved_command + "' is not of type '" + command_type + "'.");
+	} else if (command_type == "NotificationCommand") {
+		if (!NotificationCommand::GetByName(resolved_command))
+			return ApiActions::CreateResult(400, "Command '" + resolved_command + "' is not of type '" + command_type + "'.");
+	}
```

Comment:
You're saying like there is a such command, but it's not of type X. Instead behave...

Reply:
And the status code?